pkg/config/config.go

```diff
@@ -19,38 +20,59 @@ type Config struct {
 		Git GitRepositories `glu:"git"`
 		OCI OCIRepositories `glu:"oci"`
 	} `glu:"sources"`
+	Server Server `glu:"server"`
 }
 
-func (c *Config) setDefaults() error {
-	if err := c.Log.setDefaults(); err != nil {
-		return err
-	}
-
-	if err := c.Sources.Git.setDefaults(); err != nil {
-		return err
-	}
+type defaulter interface {
+	setDefaults() error
+}
 
-	if err := c.Sources.OCI.setDefaults(); err != nil {
-		return err
-	}
+func (c *Config) SetDefaults() error {
+	return processStruct(reflect.ValueOf(c).Elem(), func(d defaulter) error {
+		return d.setDefaults()
+	})
+}
 
-	return nil
+type validater interface {
+	validate() error
 }
 
 func (c *Config) Validate() error {
-	if err := c.Log.validate(); err != nil {
-		return err
-	}
+	return processStruct(reflect.ValueOf(c).Elem(), func(v validater) error {
+		return v.validate()
+	})
+}
 
-	if err := c.Sources.Git.validate(); err != nil {
-		return err
-	}
+func processStruct[T any](val reflect.Value, method func(T) error) error {
+	switch val.Kind() {
+	case reflect.Struct:
+		// Need to get addressable value for pointer receiver methods
+		if val.CanAddr() {
+			if impl, ok := val.Addr().Interface().(T); ok {
+				if err := method(impl); err != nil {
+					return err
+				}
+			}
+		}
 
-	if err := c.Sources.OCI.validate(); err != nil {
-		return err
+		// Recursively process only struct fields
+		for i := 0; i < val.NumField(); i++ {
+			field := val.Field(i)
+			if field.Kind() == reflect.Struct ||
+				(field.Kind() == reflect.Ptr && !field.IsNil() && field.Elem().Kind() == reflect.Struct) {
+				if err := processStruct(field, method); err != nil {
+					return err
+				}
+			}
+		}
+
+	case reflect.Ptr:
+		if !val.IsNil() && val.Elem().Kind() == reflect.Struct {
+			return processStruct(val.Elem(), method)
+		}
```

Comment:
We need to add a condition for `reflect.Map` kind.
SInce our maps already implement `defaulter` and `validator` we can just cast to the interface type and call the method:
```suggestion
		}
	case reflect.Map:
		if !val.IsNil() {
			if impl, ok := val.Addr().Interface().(T); ok {
				if err := method(impl); err != nil {
					return err
				}
			}
		}
```

Alternatively, we could remove the defaulter and validator implementations from the maps and just have it on the map values. Then use reflection to loop over these values without having to define this on map types. The only downside is this might not be tractable if we want to add e.g. `Name string` onto the child elements struct in the future and set the map elementrs name to match their key in the map. But it has a nice easy to not forget to implement it on the map types tradeoff:
```go
		if !val.IsNil() {
			for iter := val.MapRange(); iter.Next(); {
				if err := processStruct(iter.Value(), method); err != nil {
					return err
				}
			}
		}
```